### ISSUE TITLE
Blurry text rendering on macOS Monterey

### ISSUE THREAD

**Issue by NoirJ0e** [rohit-px2/nvui#87]
I follow the instructions wrote in the wiki and build it on my own machine, the output in the terminal seems ok 

<img width="1058" alt="image" src="https://user-images.githubusercontent.com/60462008/139446843-27c40695-f152-44ab-a0ca-324f737e3f55.png">

except the warnings but I believe it's because the system version is too new

but the program's appearance isn't good, the whole UI seems runs in low resolution 

<img width="612" alt="image" src="https://user-images.githubusercontent.com/60462008/139447254-f97ce5df-151d-405b-8817-71cb388fd4ec.png">

and I don't know if it's the way I use it wrong or something, what I expected to see is use `nvui filename` and it will open up a instance and let me modify the file, but what happened is it just create a instance but doesn't open the file I want, I have to drag the file into the instance or open it in the instance by some plugin's help

<img width="153" alt="image" src="https://user-images.githubusercontent.com/60462008/139448083-2794560f-84dd-487e-8830-a1a6fdac0bc2.png">
<img width="154" alt="image" src="https://user-images.githubusercontent.com/60462008/139448099-e62464fa-3ee2-40a1-9718-fff7d75f5955.png">

**Comment by NoirJ0e** [comment]
update: I add the `--ext_cmdline=true` and surprisingly the text in the external block has been rendered correctly 
<img width="239" alt="image" src="https://user-images.githubusercontent.com/60462008/139569640-0110a6aa-a1cb-47e3-8e48-cbed41af038e.png">

**Comment by rohit-px2** [comment]
It's probably because of the different text rendering mechanisms. Cmdline (and popup menu) uses `QPainter::drawText` since I am not doing any text caching there (not much text to draw). Meanwhile on the main text grids the text is being created as `QStaticText`, stored, and then drawn using `QPainter::drawStaticText`. I'm guessing that there's some issues with drawStaticText on mac. I'll let you know when I've added a fix.

**Comment by rohit-px2** [comment]
Does #89 fix this?

**Comment by NoirJ0e** [comment]
Sadly no :(
<img width="612" alt="image" src="https://user-images.githubusercontent.com/60462008/139614221-272c1d88-a764-471c-97d0-c37d1b17dc58.png">

**Comment by rohit-px2** [comment]
Rip. I don't really know what's causing the issue then. I'll tag this as 'help wanted', maybe someone else knows why this is happening.
Maybe you could try other fonts, do they have the same problem?

**Comment by ohmree** [comment]
I'm getting something similar on Arch Linux, X11, KDE Plasma, doom-nvim and [the `nvui` package from the AUR](https://aur.archlinux.org/packages/nvui/) started like this `/usr/share/nvui/bin/nvui --ext_multigrid=true --ext_popupmenu=true --ext_cmdline=true`:

![image](https://user-images.githubusercontent.com/13455401/141345313-2497c45f-8c1f-45a9-8a20-5ac51521675e.png)

I think my font configuration is fine since it works fine in doom emacs, kitty and a few other apps.

Maybe it has to do with display scaling? I have mine set to 125% in KDE's system settings.
I've tried setting it to 100% and restarting nvui with no success, and I'll try fully restarting the graphical session later to see if that fixes it.

**Comment by rohit-px2** [comment]
I added some commits for this today (setting antialiasing flags for rendering) on #89, could you check it out and see if it fixes your problem? Builds are [here](https://github.com/rohit-px2/nvui/actions/runs/1450028450).

**Comment by NoirJ0e** [comment]
> I'm getting something similar on Arch Linux, X11, KDE Plasma, doom-nvim and [the `nvui` package from the AUR](https://aur.archlinux.org/packages/nvui/) started like this `/usr/share/nvui/bin/nvui --ext_multigrid=true --ext_popupmenu=true --ext_cmdline=true`:
> 
> ![image](https://user-images.githubusercontent.com/13455401/141345313-2497c45f-8c1f-45a9-8a20-5ac51521675e.png)
> 
> I think my font configuration is fine since it works fine in doom emacs, kitty and a few other apps.
> 
> Maybe it has to do with display scaling? I have mine set to 125% in KDE's system settings. I've tried setting it to 100% and restarting nvui with no success, and I'll try fully restarting the graphical session later to see if that fixes it.

I used to have the same problem as you but I can't represent it every time. My workaround solution is reboot the machine and it will work for me until the next reboot, quite strange.

**Comment by NoirJ0e** [comment]
> I added some commits for this today (setting antialiasing flags for rendering) on #89, could you check it out and see if it fixes your problem? Builds are [here](https://github.com/rohit-px2/nvui/actions/runs/1450028450).

<img width="612" alt="image" src="https://user-images.githubusercontent.com/60462008/141414397-f066d6a2-5aaf-489a-8fba-a0c90eec89cc.png">
Still have the same issue on macOS, maybe this is some QT problem on Monterey, we may need to wait until QT release new version for Monterey

**Comment by kaipingdalao** [comment]
cmake --build build --target nvui --config Release
fatal error: unable to execute command: Segmentation fault: 11
in Monterey,building error,please help me

**Comment by tthkbw** [comment]
As much as I would like to play with nvui, is it unusable on macOS Monterey because of this font rendering issue. I have tried multiple fonts and sizes--some are better that others, but they all give me a headache. I feel like I'm editing on an old CRT.

Apparently no progress on this recently? 

I note that some have commented this could be a QT problem on Monterey. This seems unlikely to me. For example, Calibre uses QT and renders cleanly.

**Comment by reykjalin** [comment]
I'm pretty sure the solution to this would be to build a macOS application bundle and set the `NSHighResolutionCapable` property to `true` in the `Info.plist` file.

[I've run into this before](https://www.thorlaksson.com/how-to-fix-blurry-text-in-macos-applications/) when playing around with Qt and other GUI frameworks and this fixed the problem for me every time.

The Qt docs have [a guide on building macOS app bundles](https://doc.qt.io/qt-5/macos-deployment.html) but it seems to only have a guide for QMake and not CMake.

The general lack of application bundles is honestly one of the things that makes me not want to invest time in setting up some of the Neovim GUIs out there because I dislike leaving my terminal open, so this could be a relatively easy win; generate an application bundle when building on macos.

**Comment by rohit-px2** [comment]
Cool, that might be what I was looking for! I'll create a new branch with the plist soon, will post it here when it's been created.

**Comment by rohit-px2** [comment]
Had to fix some build issues so it took me longer than I thought. I added the plist on the `blurry_text_fix_mac` branch and you can find builds [here](https://github.com/rohit-px2/nvui/actions/runs/1800036151), let me know if that solves it.

**Comment by tthkbw** [comment]
I downloaded the build for macOS and tried it. It doesn't seem to have any effect on the blurry text problem on my machine--m1 mini running the latest Monterey, 12.2 with a LG 5k monitor.

It is not obvious to me that the plist file is being used. The structure of the app and location of the plist is the same as other .app packages I looked at. I also moved the nvui.app to the /Applications directory. However, I can only run nvui.app successfully by using the terminal `open /Applications/nvui.app` command. Running from Finder causes nvui to report that there is no Neovim in the PATH. I'm not sure what Finder thinks $PATH might be. 

I don't know how to determine if nvui.app is actually using the info.plist included with it. If it is, then the problem is not fixed. If it is not, how to I make it so?

**Comment by rohit-px2** [comment]
I used CMake's bundle feature to bundle the plist with the app so it should be using the plist automatically. I don't have any other ideas besides that. Could you check if Neovim-Qt/Goneovim have the same rendering issue? They also render using Qt so if they are fine then there could be a way they solved it.
As for the PATH issue, no idea on why that's happening.

**Comment by tthkbw** [comment]
I loaded Goneovim version 0.5.1. It works and renders perfectly. Goneovim's plist file has quite a bit more 'stuff' than nvui's does. nvui has:

`<dict>
      <key>NSPrincipalClass</key>
      <string>NSApplication</string>
      <key>NSHighResolutionCapable</key>
      <true/>
    </dict>`

for the same key, Goneovim has:

`	<key>NSPrincipalClass</key>
	<string>NSApplication</string>'

but Goneovim does not have the key NSHighResolutionCapable.

**Comment by tthkbw** [comment]
I’ll watch for the build and test for you.

**Comment by rohit-px2** [comment]
Tried the build myself (on Windows) and it looks like the automatic high dpi scaling has broken a lot of things.. I'll try and fix them and put a comment here when it's done

**Comment by rohit-px2** [comment]
@tthkbw I think the scaling issues have been fixed so if you could let me know how it's working that would be great. The build is [here](https://github.com/rohit-px2/nvui/actions/runs/1837488167)

**Comment by tthkbw** [comment]
Yes, it is fixed and displays properly now. However, the problem with running nvui.app from anywhere other than the command line using `/Applications/nvui.app` persists. Nvui must be using the PATH environment variable to find nvim, but I don't think the PATH environment variable exists other than from the terminal.

**Comment by rohit-px2** [comment]
Nice! And yeah, from what I've seen in Goneovim and Neovim-Qt, they spawn an external process to print the $PATH and then they manually set it. Might have to do something similar

**Comment by reykjalin** [comment]
> I think the scaling issues have been fixed so if you could let me know how it's working that would be great. The build is [here](https://github.com/rohit-px2/nvui/actions/runs/1837488167)

Hmm, when I try to run this I see the following:

```
$ open ~/Desktop/nvui-macos-x64/bin/nvui.app/

The application cannot be opened for an unexpected reason, error=Error Domain=RBSRequestErrorDomain Code=5 "Launch failed." UserInfo={NSLocalizedFailureReason=Launch failed., NSUnderlyingError=0x600003611470 {Error Domain=NSPOSIXErrorDomain Code=111 "Unknown error: 111" UserInfo={NSLocalizedDescription=Launchd job spawn failed}}}
```

I wonder if that's because the CI is building that on intel architecture while I'm using the M1 chip 🤔

**Comment by rohit-px2** [comment]
Yeah only the x64 version is built, Github Actions doesn't have M1 virtual machines so that's probably why.

**Comment by coordinatio** [comment]
I observe very similar issue on Arch Linux. Here is how font rendering inside the latest nvui looks like in comparison to fonts in nvim-qt:
![image](https://user-images.githubusercontent.com/67314678/168383780-7fb98086-89f5-4a3c-8944-b1da4fda334d.png)
 
Please tell me if you know what information needed to help investigate the issue. I will try to provide it.

**Comment by coordinatio** [comment]
I have discovered that the issue goes away if the display scaling is switched off. Maybe that can help to resolve the issue. 

I use Xorg and KDE and set global display scaling through KDE's systemsettings utility.
